File: tests/stringUtils.test.js
const { capitalizeWords } = require('../src/stringUtils');

describe('capitalizeWords', () => {
  describe('Standard functionality', () => {
    test('should capitalize first letter of each word in a simple sentence', () => {
      expect(capitalizeWords('hello world')).toBe('Hello World');
    });

    test('should handle empty string', () => {
      expect(capitalizeWords('')).toBe('');
    });

    test('should handle single word', () => {
      expect(capitalizeWords('hello')).toBe('Hello');
    });

    test('should handle already capitalized words', () => {
      expect(capitalizeWords('Hello World')).toBe('Hello World');
    });

    test('should handle mixed case words', () => {
      expect(capitalizeWords('hELLo wORLd')).toBe('HELLo WORLd');
    });

    test('should preserve multiple spaces between words', () => {
      expect(capitalizeWords('hello  world')).toBe('Hello  World');
      expect(capitalizeWords('hello   world   test')).toBe('Hello   World   Test');
    });

    test('should handle leading and trailing spaces', () => {
      expect(capitalizeWords('  hello world  ')).toBe('  Hello World  ');
    });

    test('should handle tabs and mixed whitespace', () => {
      expect(capitalizeWords('hello\tworld')).toBe('Hello\tWorld');
      expect(capitalizeWords('hello \t world')).toBe('Hello \t World');
    });
  });

  describe('Punctuation and special cases', () => {
    test('should handle punctuation', () => {
      expect(capitalizeWords('hello, world!')).toBe('Hello, World!');
      expect(capitalizeWords('hello; world?')).toBe('Hello; World?');
    });

    test('should handle apostrophes correctly', () => {
      expect(capitalizeWords("don't stop")).toBe("Don't Stop");
      expect(capitalizeWords("it's amazing")).toBe("It's Amazing");
    });

    test('should handle hyphens', () => {
      expect(capitalizeWords('state-of-the-art')).toBe('State-of-the-art');
      expect(capitalizeWords('well-known example')).toBe('Well-known Example');
    });

    test('should handle numbers', () => {
      expect(capitalizeWords('test 123 abc')).toBe('Test 123 Abc');
      expect(capitalizeWords('room 42b floor 3')).toBe('Room 42b Floor 3');
    });

    test('should handle strings with only numbers', () => {
      expect(capitalizeWords('123 456')).toBe('123 456');
    });

    test('should handle strings with special characters', () => {
      expect(capitalizeWords('hello@world.com')).toBe('Hello@world.com');
      expect(capitalizeWords('price: $100')).toBe('Price: $100');
    });
  });

  describe('Unicode and international characters', () => {
    test('should handle accented characters', () => {
      expect(capitalizeWords('café naïve')).toBe('Café Naïve');
      expect(capitalizeWords('résumé über')).toBe('Résumé Über');
    });

    test('should handle non-Latin scripts', () => {
      expect(capitalizeWords('hello мир')).toBe('Hello Мир');
    });

    test('should handle mixed scripts', () => {
      expect(capitalizeWords('test 测试')).toBe('Test 测试');
    });
  });

  describe('Input validation and error handling', () => {
    test('should handle null input', () => {
      expect(capitalizeWords(null)).toBe('');
    });

    test('should handle undefined input', () => {
      expect(capitalizeWords(undefined)).toBe('');
    });

    test('should throw error for non-string, non-null inputs', () => {
      expect(() => capitalizeWords(123)).toThrow('Input must be a string');
      expect(() => capitalizeWords({})).toThrow('Input must be a string');
      expect(() => capitalizeWords([])).toThrow('Input must be a string');
      expect(() => capitalizeWords(true)).toThrow('Input must be a string');
    });

    test('should handle very long strings', () => {
      const longString = 'word '.repeat(10000).trim();
      const result = capitalizeWords(longString);
      expect(result).toMatch(/^Word( Word)*$/);
      expect(result.split(' ').length).toBe(10000);
    });

    test('should reject strings exceeding maximum length', () => {
      const maxLength = 10 * 1024 * 1024; // 10MB
      const tooLongString = 'a'.repeat(maxLength + 1);
      expect(() => capitalizeWords(tooLongString)).toThrow('Input string exceeds maximum length');
    });
  });

  describe('Security tests', () => {
    test('should handle script tags without executing', () => {
      const maliciousInput = '<script>alert("xss")</script>';
      const result = capitalizeWords(maliciousInput);
      expect(result).toBe('<script>Alert("xss")</script>');
      expect(typeof result).toBe('string');
    });

    test('should handle SQL injection patterns', () => {
      const sqlInjection = "'; DROP TABLE users; --";
      const result = capitalizeWords(sqlInjection);
      expect(result).toBe("'; DROP TABLE Users; --");
      expect(typeof result).toBe('string');
    });

    test('should handle null byte injection', () => {
      const nullByteInput = 'hello\0world';
      const result = capitalizeWords(nullByteInput);
      expect(result).toMatch(/^Hello\0World$/);
    });

    test('should handle control characters safely', () => {
      const controlChars = 'hello\x01\x02world';
      expect(() => {
        const result = capitalizeWords(controlChars);
        expect(typeof result).toBe('string');
      }).not.toThrow();
    });

    test('should handle newlines and carriage returns', () => {
      expect(capitalizeWords('hello\nworld')).toBe('Hello\nWorld');
      expect(capitalizeWords('hello\r\nworld')).toBe('Hello\r\nWorld');
    });

    test('should handle HTML entities', () => {
      const htmlEntities = '&lt;hello&gt; &amp;world';
      const result = capitalizeWords(htmlEntities);
      expect(result).toBe('&lt;Hello&gt; &amp;world');
    });

    test('should not evaluate code-like strings', () => {
      const codeString = 'eval("malicious code")';
      const result = capitalizeWords(codeString);
      expect(result).toBe('Eval("malicious Code")');
      expect(typeof result).toBe('string');
    });
  });

  describe('Edge cases', () => {
    test('should handle single character', () => {
      expect(capitalizeWords('a')).toBe('A');
      expect(capitalizeWords('z')).toBe('Z');
    });

    test('should handle only whitespace', () => {
      expect(capitalizeWords('   ')).toBe('   ');
      expect(capitalizeWords('\t\t')).toBe('\t\t');
    });

    test('should handle only punctuation', () => {
      expect(capitalizeWords('!!!')).toBe('!!!');
      expect(capitalizeWords('...')).toBe('...');
    });

    test('should handle emoji', () => {
      expect(capitalizeWords('hello 😊 world')).toBe('Hello 😊 World');
    });

    test('should handle repeated punctuation', () => {
      expect(capitalizeWords('hello!!! world???')).toBe('Hello!!! World???');
    });

    test('should preserve original string type', () => {
      const result = capitalizeWords('test');
      expect(typeof result).toBe('string');
      expect(result).not.toBe(null);
      expect(result).not.toBe(undefined);
    });
  });

  describe('Performance tests', () => {
    test('should handle large strings efficiently', () => {
      const largeString = 'word '.repeat(50000).trim();
      const startTime = Date.now();
      const result = capitalizeWords(largeString);
      const endTime = Date.now();
      
      expect(result).toBeDefined();
      expect(endTime - startTime).toBeLessThan(5000); // Should complete within 5 seconds
    });

    test('should handle many small words', () => {
      const manyWords = Array.from({ length: 10000 }, (_, i) => `w${i}`).join(' ');
      const result = capitalizeWords(manyWords);
      expect(result).toMatch(/^W0 W1/);
    });
  });
});

File: src/test/java/com/project/utils/StringUtilsTest.java
package com.project.utils;

import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.params.ParameterizedTest;
import org.junit.jupiter.params.provider.CsvSource;
import org.junit.jupiter.params.provider.NullAndEmptySource;
import org.junit.jupiter.params.provider.ValueSource;

import static org.junit.jupiter.api.Assertions.*;

/**
 * Comprehensive unit tests for StringUtils.capitalizeWords functionality.
 * Tests cover standard cases, edge cases, security scenarios, and performance.
 */
class StringUtilsTest {

    @Test
    @DisplayName("Should capitalize first letter of each word in standard case")
    void testCapitalizeWords_StandardCase() {
        String input = "hello world";
        String expected = "Hello World";
        assertEquals(expected, StringUtils.capitalizeWords(input));
    }

    @Test
    @DisplayName("Should handle null input gracefully")
    void testCapitalizeWords_NullInput() {
        assertThrows(IllegalArgumentException.class, () -> {
            StringUtils.capitalizeWords(null);
        });
    }

    @Test
    @DisplayName("Should handle empty string")
    void testCapitalizeWords_EmptyString() {
        String input = "";
        String expected = "";
        assertEquals(expected, StringUtils.capitalizeWords(input));
    }

    @Test
    @DisplayName("Should handle single word")
    void testCapitalizeWords_SingleWord() {
        String input = "hello";
        String expected = "Hello";
        assertEquals(expected, StringUtils.capitalizeWords(input));
    }

    @Test
    @DisplayName("Should preserve multiple spaces between words")
    void testCapitalizeWords_MultipleSpaces() {
        String input = "hello  world";
        String expected = "Hello  World";
        assertEquals(expected, StringUtils.capitalizeWords(input));
    }

    @Test
    @DisplayName("Should handle punctuation correctly")
    void testCapitalizeWords_WithPunctuation() {
        String input = "hello, world!";
        String expected = "Hello, World!";
        assertEquals(expected, StringUtils.capitalizeWords(input));
    }

    @Test
    @DisplayName("Should handle numbers in text")
    void testCapitalizeWords_WithNumbers() {
        String input = "test 123 abc";
        String expected = "Test 123 Abc";
        assertEquals(expected, StringUtils.capitalizeWords(input));
    }

    @Test
    @DisplayName("Should handle Unicode characters properly")
    void testCapitalizeWords_Unicode() {
        String input = "café naïve";
        String expected = "Café Naïve";
        assertEquals(expected, StringUtils.capitalizeWords(input));
    }

    @Test
    @DisplayName("Should handle apostrophes in contractions")
    void testCapitalizeWords_Apostrophes() {
        String input = "don't can't won't";
        String expected = "Don't Can't Won't";
        assertEquals(expected, StringUtils.capitalizeWords(input));
    }

    @Test
    @DisplayName("Should handle leading and trailing spaces")
    void testCapitalizeWords_LeadingTrailingSpaces() {
        String input = "  hello world  ";
        String expected = "  Hello World  ";
        assertEquals(expected, StringUtils.capitalizeWords(input));
    }

    @Test
    @DisplayName("Should handle only spaces")
    void testCapitalizeWords_OnlySpaces() {
        String input = "     ";
        String expected = "     ";
        assertEquals(expected, StringUtils.capitalizeWords(input));
    }

    @Test
    @DisplayName("Should handle mixed case input")
    void testCapitalizeWords_MixedCase() {
        String input = "hELLo WoRLd";
        String expected = "HELLo WoRLd";
        assertEquals(expected, StringUtils.capitalizeWords(input));
    }

    @Test
    @DisplayName("Should handle newlines and tabs")
    void testCapitalizeWords_NewlinesAndTabs() {
        String input = "hello\nworld\ttest";
        String expected = "Hello\nWorld\tTest";
        assertEquals(expected, StringUtils.capitalizeWords(input));
    }

    @ParameterizedTest
    @CsvSource({
        "'a b c', 'A B C'",
        "'the quick brown fox', 'The Quick Brown Fox'",
        "'one two three four five', 'One Two Three Four Five'",
        "'test-case', 'Test-case'",
        "'hello.world', 'Hello.world'"
    })
    @DisplayName("Should handle various input patterns")
    void testCapitalizeWords_VariousPatterns(String input, String expected) {
        assertEquals(expected, StringUtils.capitalizeWords(input));
    }

    @Test
    @DisplayName("Should handle already capitalized text")
    void testCapitalizeWords_AlreadyCapitalized() {
        String input = "Hello World";
        String expected = "Hello World";
        assertEquals(expected, StringUtils.capitalizeWords(input));
    }

    @Test
    @DisplayName("Should handle single character words")
    void testCapitalizeWords_SingleCharWords() {
        String input = "a b c d e";
        String expected = "A B C D E";
        assertEquals(expected, StringUtils.capitalizeWords(input));
    }

    @Test
    @DisplayName("Should handle special characters and symbols")
    void testCapitalizeWords_SpecialCharacters() {
        String input = "hello @world #test $money";
        String expected = "Hello @world #test $money";
        assertEquals(expected, StringUtils.capitalizeWords(input));
    }

    // Security-focused tests

    @Test
    @DisplayName("Security: Should handle script tags without execution")
    void testCapitalizeWords_SecurityScriptTags() {
        String input = "<script>alert('xss')</script>";
        String result = StringUtils.capitalizeWords(input);
        assertNotNull(result);
        assertFalse(result.contains("script>alert"));
        // Result should be capitalized but no execution
        assertTrue(result.startsWith("<script>"));
    }

    @Test
    @DisplayName("Security: Should handle SQL injection patterns")
    void testCapitalizeWords_SecuritySQLInjection() {
        String input = "'; DROP TABLE users; --";
        String result = StringUtils.capitalizeWords(input);
        assertNotNull(result);
        assertTrue(result.contains("DROP"));
        // Should just capitalize, not execute
    }

    @Test
    @DisplayName("Security: Should handle null byte injection attempts")
    void testCapitalizeWords_SecurityNullByte() {
        String input = "hello\0world";
        String result = StringUtils.capitalizeWords(input);
        assertNotNull(result);
    }

    @Test
    @DisplayName("Security: Should reject extremely large strings")
    void testCapitalizeWords_SecurityDOSPrevention() {
        // Create string larger than typical limit (e.g., 10MB)
        int size = 11 * 1024 * 1024; // 11 MB
        StringBuilder sb = new StringBuilder(size);
        for (int i = 0; i < size; i++) {
            sb.append('a');
        }
        String input = sb.toString();
        
        assertThrows(IllegalArgumentException.class, () -> {
            StringUtils.capitalizeWords(input);
        });
    }

    @Test
    @DisplayName("Security: Should handle control characters safely")
    void testCapitalizeWords_SecurityControlCharacters() {
        String input = "hello\u0001world\u0002test";
        String result = StringUtils.capitalizeWords(input);
        assertNotNull(result);
    }

    @Test
    @DisplayName("Security: Should handle Unicode direction override characters")
    void testCapitalizeWords_SecurityUnicodeOverride() {
        String input = "hello\u202Eworld";
        String result = StringUtils.capitalizeWords(input);
        assertNotNull(result);
    }

    // Performance tests

    @Test
    @DisplayName("Performance: Should handle moderately large strings efficiently")
    void testCapitalizeWords_PerformanceModerate() {
        StringBuilder sb = new StringBuilder();
        for (int i = 0; i < 10000; i++) {
            sb.append("word ");
        }
        String input = sb.toString().trim();
        
        long startTime = System.currentTimeMillis();
        String result = StringUtils.capitalizeWords(input);
        long endTime = System.currentTimeMillis();
        
        assertNotNull(result);
        assertTrue(result.startsWith("Word"));
        assertTrue((endTime - startTime) < 5000, "Operation took too long");
    }

    @Test
    @DisplayName("Should handle strings with only special characters")
    void testCapitalizeWords_OnlySpecialChars() {
        String input = "!@#$%^&*()";
        String expected = "!@#$%^&*()";
        assertEquals(expected, StringUtils.capitalizeWords(input));
    }

    @Test
    @DisplayName("Should handle strings with repeated delimiters")
    void testCapitalizeWords_RepeatedDelimiters() {
        String input = "hello...world";
        String expected = "Hello...world";
        assertEquals(expected, StringUtils.capitalizeWords(input));
    }

    @Test
    @DisplayName("Should handle accented characters")
    void testCapitalizeWords_AccentedCharacters() {
        String input = "éclair résumé";
        String expected = "Éclair Résumé";
        assertEquals(expected, StringUtils.capitalizeWords(input));
    }

    @Test
    @DisplayName("Should handle hyphenated words")
    void testCapitalizeWords_HyphenatedWords() {
        String input = "well-known state-of-the-art";
        String expected = "Well-known State-of-the-art";
        assertEquals(expected, StringUtils.capitalizeWords(input));
    }

    @Test
    @DisplayName("Should handle all uppercase input")
    void testCapitalizeWords_AllUppercase() {
        String input = "HELLO WORLD";
        String expected = "HELLO WORLD";
        assertEquals(expected, StringUtils.capitalizeWords(input));
    }

    @Test
    @DisplayName("Should handle all lowercase input")
    void testCapitalizeWords_AllLowercase() {
        String input = "hello world test";
        String expected = "Hello World Test";
        assertEquals(expected, StringUtils.capitalizeWords(input));
    }

    @Test
    @DisplayName("Should handle words starting with numbers")
    void testCapitalizeWords_WordsStartingWithNumbers() {
        String input = "123abc 456def";
        String expected = "123abc 456def";
        assertEquals(expected, StringUtils.capitalizeWords(input));
    }

    @Test
    @DisplayName("Should handle whitespace string")
    void testCapitalizeWords_WhitespaceString() {
        String input = "\t\n\r ";
        String result = StringUtils.capitalizeWords(input);
        assertNotNull(result);
    }
}